src/pip/_internal/req/req_install.py

```diff
@@ -725,13 +770,23 @@ def metadata(self):
         return self._metadata
 
     def get_dist(self):
-        """Return a pkg_resources.Distribution built from self.egg_info_path"""
-        egg_info = self.egg_info_path.rstrip(os.path.sep)
-        base_dir = os.path.dirname(egg_info)
-        metadata = pkg_resources.PathMetadata(base_dir, egg_info)
-        dist_name = os.path.splitext(os.path.basename(egg_info))[0]
-        return pkg_resources.Distribution(
-            os.path.dirname(egg_info),
+        """Return a pkg_resources.Distribution for this requirement"""
+        if self.metadata_directory:
```

Comment:
Or even better:
```python
    def get_dist(self):
        """Return a pkg_resources.Distribution for this requirement"""
        if self.metadata_directory:
            metadata_path = self.metadata_directory
        else:
            metadata_path = self.egg_info_path
        metadata_path = metadata_path.rstrip(os.path.sep)
        return next(pkg_resources.distributions_from_metadata(metadata_path))
```

Reply:
I'm OK with this - but don't really know the pkg_resources functions well enough to judge the impact. I'll wait for @pradyunsg to do his review, and then I'm happy to just replace the existing `get_dist` with this one.

On the other hand, it would also be just as easy to put in what we have, and then follow up with a PR simplifying `get_dist` (in the interest of smaller more focused PRs being better). But as long as making the change doesn't trigger a new wave of weird CI failures, it doesn't really make much difference - this PR is already way too complicated, one more commit won't make much difference 😉

Reply:
I'm on board for making more follow up PRs for cleanups. This PR is already a little unwieldy in size and it's taken me a lot longer than I estimated to even try it out.

I've bumped existing notes just so that I don't have to scroll too high when taking a pass later to see what places we've already identified for improvements/refactors.